File: hoerbuchkatalog/domain/src/main/java/wbh/bookworm/hoerbuchkatalog/domain/Hoerbuchkatalog.java
/*
 * Copyright (C) 2011-2018 art of coding UG, https://www.art-of-coding.eu
 * Alle Rechte vorbehalten. Nutzung unterliegt Lizenzbedingungen.
 * All rights reserved. Use is subject to license terms.
 */

package wbh.bookworm.hoerbuchkatalog.domain;

import java.util.Map;
import java.util.Set;
import java.util.TreeSet;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

/**
 * Aggregate
 */
public final class Hoerbuchkatalog extends DddAggregate<Map<Titelnummer, Hoerbuch>> {

    private final Map<Titelnummer, Hoerbuch> katalog;

    public Hoerbuchkatalog() {
        this.katalog = new ConcurrentHashMap<>();
    }

    public Hoerbuchkatalog(final Map<Titelnummer, Hoerbuch> katalog) {
        this.katalog = katalog;
    }

    public void hinzufuegen(final Hoerbuch hoerbuch) {
        katalog.put(hoerbuch.getTitelnummer(), hoerbuch);
    }

    public int anzahlHoerbuecher() {
        return katalog.size();
    }

    public int anzahlDownloadbarerHoerbuecher() {
        return katalog.values().stream()
                .filter(Hoerbuch::isDownloadbar)
                .collect(Collectors.toUnmodifiableList())
                .size();
    }

    public Set<Hoerbuch> alleHoerbuecher() {
        final TreeSet<Hoerbuch> treeSet = new TreeSet<>(Hoerbuch::compareTo);
        treeSet.addAll(katalog.values());
        return treeSet;
    }

    public Hoerbuch hoerbuch(final Titelnummer titelnummer) {
        return katalog.get(titelnummer);
    }

    @Override
    public int compareTo(final Map<Titelnummer, Hoerbuch> o) {
        return 0;
    }

}

File: hoerbuchkatalog/webservice/src/main/resources/static/js/hoerbuchkatalog.js
/*
 * Copyright (C) 2020 art of coding UG, https://www.art-of-coding.eu
 * Alle Rechte vorbehalten. Nutzung unterliegt Lizenzbedingungen.
 * All rights reserved. Use is subject to license terms.
 */

"use strict";

import {Bookworm} from "./lib/bookworm.js";

window.onload = function () {
    const bookworm = new Bookworm();
    // Hörprobe
    const hoerprobeButton = document.querySelector('.button .hoerprobe');
    hoerprobeButton.addEventListener('click', () => {
        const titelnummer = document.querySelector('#titelnummer').value;
        bookworm.hoerprobe(titelnummer);
    });
    // Merkliste
    const merklisteButton = document.querySelector('.button .watchlist-false');
    merklisteButton.addEventListener('click', function (e) {
        console.log(e);
        console.log(this);
    });
    // CD Bestellung
    const cdBestellenButton = document.querySelector('.button .order-cd-false');
    cdBestellenButton.addEventListener('click', function (e) {
        console.log(e);
        console.log(this);
    });
    // Download
    const downloadButton = document.querySelector('.button .order-download-true');
    downloadButton.addEventListener('click', () => {
    });
};
